### PR TITLE
[Feat] Refactor of vizro_ai.plot (breaking)

### PR DESCRIPTION
**Pull request by maxschulz-COL**
## Description
Full refactor of `vizro-ai.plot`. For reviewers (before docs are online):

New API (no more explain, but added `validate_code`):
```
df: pd.DataFrame,
user_input: str,
max_debug_retry: int = 3,
return_elements: bool = False,
validate_code: bool = True
```

Cases are below:

`return_elements = False` - return Vizro chart, but no explanation, or code explanation

`return_elements = True` - return response pydantic model. This model has the following properties and methods:

- `response.code` - get dash code
- `response.code_vizro` - get Vizro code
- `response.get_fig_object(self, data_frame: pd.DataFrame, chart_name: Optional[str] = None, vizro=True)`
--> re-execute code to obtain fig object with DF of choice, with a different chart name, and as Vizro or plotly chart

`validate_code: bool = False` - does not execute any code, but code string is not validated by actually running it

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-ai/src/vizro_ai/plot2/_response_models.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    # query = "plot a bubble chart to shows the changes in life expectancy gdp per capita  over time. Animate the chart by year"
+
+    # print(df.sample(10).to_string())
+
+    # res = _get_pydantic_model(query=query, llm_model=model, response_model=dummy_model, df_info=df.sample(10).to_string())
+    # code = res._get_complete_code(lint=True)
+    # fig = res._get_fig_object(data_frame=df)
+    # fig.show()
+
+    # llm = ChatAnthropic(
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/plot2/_response_models.py:204]: btw, I tested with Anthropic and it works very well 👏 you just need to ensure the base url is correct. One of these 3 did the trick:
```
print(os.environ.get("ANTHROPIC_API_URL"))
print(os.environ.get("ANTHROPIC_BASE_URL"))
print(os.environ.get("ANTHROPIC_API_BASE"))
```

**Review by lingyielia — COMMENTED** [review]
Very excited to seeing this PR now shaping up!

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 2]
```diff
@@ -160,6 +162,41 @@ def plot(
...
         return vizro_plot if return_elements else vizro_plot.figure
 
+    def plot2(
+        self,
+        df: pd.DataFrame,
+        user_input: str,
+        max_debug_retry: int = 3,
+        return_elements: bool = False,
+        validate_code: bool = True,
+    ) -> Union[go.Figure, ChartPlanStatic]:
```
[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py:172]: Can definitely change - what would suggest?

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 3]
```diff
@@ -160,6 +162,41 @@ def plot(
 
         return vizro_plot if return_elements else vizro_plot.figure
 
+    def plot2(
```
[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py:165]:
```py
res = vizro_ai.plot2()
res.get_fig_object(chart_name = xxx)
```
or similar with the `_get_complete_code` function

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 4]
```diff
@@ -160,6 +162,41 @@ def plot(
...
+            max_debug_retry: Maximum number of retries to debug errors. Defaults to `3`.
+            return_elements: Flag to return ChartPlanStatic pydantic model that includes all possible elements generated.
+                Defaults to `False`.
+            validate_code: Flag if produced code should be executed to validate it. Defaults to `True`.
+
+        Returns:
+            go.Figure or ChartPlanStatic pydantic model
+
+        """
+        response_model = ChartPlanDynamicFactory(data_frame=df) if validate_code else ChartPlanStatic
```
[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py:187]: Hmm, not sure I like it. Could be rather do that at figure build time?

Something like:
```py
def get_fig_object(self, data_frame: Union[pd.Dataframe,str], chart_name: Optional[str] = None, vizro=True):
        """Get the figure object."""
        ldict = {}
        chart_name = chart_name or CUSTOM_CHART_NAME
        code_to_execute = self._get_complete_code(chart_name=chart_name, vizro=vizro)
        exec(code_to_execute, globals(), ldict)
        chart = ldict[f"{chart_name}"]
        return chart(data_frame)
```

File: vizro-ai/src/vizro_ai/plot2/_response_models.py (hunk 1 above)
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/plot2/_response_models.py:204]: I had very good experiences with anthropic, that is really good to know

File: vizro-ai/src/vizro_ai/plot/_response_models.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        return create_model(
+            "ChartPlanDynamic",
+            __validators__={
+                "validator1": validator("chart_code", allow_reuse=True)(_test_execute_chart_code),
+            },
+            __base__=ChartPlanStatic,
+        )
+
+
+if __name__ == "__main__":
```
Comment (maxschulz-COL) [inline vizro-ai/src/vizro_ai/plot/_response_models.py:189]: Delete before merge

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 6]
```diff
@@ -48,120 +38,41 @@ def __init__(self, model: Optional[Union[ChatOpenAI, str]] = None):
...
-        """
-        # TODO retained for some chat application integration, need deprecation handling
-        return self._run_plot_tasks(df, user_input, explain=False).code
 
     def plot(
         self,
         df: pd.DataFrame,
         user_input: str,
-        explain: bool = False,
         max_debug_retry: int = 3,
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/_vizro_ai.py:46]: Did you try estimate the optimal retry number? I understand retry is helpful as it can feedback the error message to llm. But wonder whether retry 3 times is more effective than 2 times? Do we observe llm converge to correct answers with more retries, or llm is directed to a direction of off-track?

Unless 3 is obviously better, I'd prefer maybe even retry once. Because it 3 means triple the latency, which is long

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 7]
```diff
@@ -48,120 +38,41 @@ def __init__(self, model: Optional[Union[ChatOpenAI, str]] = None):
...
-
-        """
-        # TODO retained for some chat application integration, need deprecation handling
-        return self._run_plot_tasks(df, user_input, explain=False).code
 
     def plot(
         self,
         df: pd.DataFrame,
         user_input: str,
-        explain: bool = False,
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/_vizro_ai.py:130]: I like the idea of removing `explain` from here, and using `return_elements` to cover it (which is much more streamlined). Although this is a small breaking change, we didn't provide any deprecation warning before fully remove it.

Just to make users' life easier, shall we maybe try improve the error message in case user uses `explain`? and provide a migration guide (or as simple as a suggested code block), so it's clear how to get explanations with new API. Something like

```
def deprecated_param_check(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        if 'explain' in kwargs:
            raise TypeError(
                "VizroAI.plot() no longer supports the 'explain' parameter. "
                "This parameter has been removed in the latest version. "
                "If you need explanations, do `return_elements=True` and `res.chart_insights` and `res.code_explanation`"
            )
        return func(*args, **kwargs)
    return wrapper
    
    
class VizroAI:
    @deprecated_param_check
    def plot()
```

Current error message is
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[8], line 1
----> 1 vizro_ai.plot(df, "describe the composition of gdp in continent, and add horizontal line for avg gdp", explain=True)

TypeError: VizroAI.plot() got an unexpected keyword argument 'explain'
```

File: vizro-ai/pyproject.toml [hunk 8]
```diff
@@ -18,10 +18,11 @@ dependencies = [
   "tabulate",
   "openai>=1.0.0",
   "langchain>=0.1.0, <0.3.0",  # TODO update all LLMChain class, update to pydantic v2 and remove upper bound
-  "langchain-openai",
   "langgraph>=0.1.2",
   "python-dotenv>=1.0.0",  # TODO decide env var management to see if we need this
-  "vizro>=0.1.20"
+  "vizro>=0.1.20",
+  "langchain_openai",  # Base dependency, ie minimum model working, but ideally we could get rid of this
+  "ruff"
```
Comment (lingyielia) [inline vizro-ai/pyproject.toml:25]: Do we want to add one more vendor in this PR? Also great if you prefer to do it in a separate PR.

```
[project.optional-dependencies]
anthropic = [
  "langchain-anthropic",
]

File: vizro-ai/src/vizro_ai/_llm_models.py [hunk 9]
```diff
@@ -21,16 +26,20 @@
         "gpt-4o-mini",
         "gpt-4o-mini-2024-07-18",
     ],
+    "Anthropic": [
+        "claude-3-sonnet-20240229",
+    ],
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/_llm_models.py:31]:
```suggestion
    "Anthropic": [
        "claude-3-sonnet-20240229",
        "claude-3-opus-20240229",
        "claude-3-haiku-20240307",
        "claude-3-5-sonnet-20240620",
    ],
```

File: vizro-ai/src/vizro_ai/dashboard/_response_models/components.py [hunk 10]
```diff
@@ -63,18 +63,20 @@ def create(self, model: BaseChatModel, all_df_metadata: AllDfMetadata) -> Compon
                 from vizro_ai import VizroAI
 
                 vizro_ai = VizroAI(model=model)
-                result = vizro_ai._run_plot_tasks(
+                result = vizro_ai.plot(
                     df=all_df_metadata.get_df(self.df_name),
                     user_input=self.component_description,
-                    chart_name=self.component_id,
-                    max_debug_retry=2,
+                    max_debug_retry=2,  # TODO must be flexible
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/dashboard/_response_models/components.py:69]: Why do we want to make it flexible? I feel this control would be too granular to be controlled by dashboard users

File: vizro-ai/src/vizro_ai/dashboard/_response_models/components.py [hunk 11]
```diff
@@ -63,18 +63,20 @@ def create(self, model: BaseChatModel, all_df_metadata: AllDfMetadata) -> Compon
...
+                    return_elements=True,
                 )
                 return ComponentResult(
                     component=vm.Graph(
                         id=self.component_id,
-                        figure=result.figure,
+                        figure=result.get_fig_object(
+                            chart_name=self.component_id, data_frame=all_df_metadata.get_df(self.df_name), vizro=True
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/dashboard/_response_models/components.py:76]: wow! 🤩

**Review by lingyielia — CHANGES_REQUESTED** [review]
Some initial feedback. Will continue tmr.

SUPER COOL PR! 👍

File: vizro-ai/src/vizro_ai/_vizro_ai.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py:46]: I did not, i left the value we had before :) I observed self correction with 1 retry, never with any more than that. We could go for 1 if you like. Let me implement that.

File: vizro-ai/src/vizro_ai/_vizro_ai.py (hunk 7 above)
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py:130]: Good idea, I like it!

Technically we could also implement this for the returned object, but that seems a little weird on the `ChartPlanStatic` no?

File: vizro-ai/src/vizro_ai/_llm_models.py (hunk 9 above)
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_llm_models.py:31]: See docs PR

File: vizro-ai/pyproject.toml (hunk 8 above)
Comment (maxschulz-COL) [reply vizro-ai/pyproject.toml:25]: See next PR - but I see the point of optional dependencies... On the docs PR I left out all but `langchain-openai`, but given that we allow for string repesentation, let me add those under optional

File: vizro-ai/src/vizro_ai/_vizro_ai.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nadijagraca) [reply vizro-ai/src/vizro_ai/_vizro_ai.py:69]: Maybe we should follow _response_models and change to `ChartPlan`?

File: vizro-ai/src/vizro_ai/dashboard/_response_models/components.py [hunk 12]
```diff
@@ -63,18 +63,20 @@ def create(self, model: BaseChatModel, all_df_metadata: AllDfMetadata) -> Compon
                 from vizro_ai import VizroAI
 
                 vizro_ai = VizroAI(model=model)
-                result = vizro_ai._run_plot_tasks(
+                result = vizro_ai.plot(
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/dashboard/_response_models/components.py:66]: We need a better fallback here if the plot failed with Validation error. Now it doesn't get caught and breaks the dashboard building process.
Screenshot 2024-09-04 at 11 31 51

File: vizro-ai/src/vizro_ai/plot/_response_models.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    )
+    chart_insights: str = Field(
+        ...,
+        description="""
+        Insights to what the chart explains or tries to show. Ideally concise and between 30 and 60 words.""",
+    )
+    code_explanation: str = Field(
+        ...,
+        description="""
+        Explanation of the code steps used for `chart_code` field. Ideally a concise bulleted step-by-step list.""",
```
Comment (lingyielia) [inline vizro-ai/src/vizro_ai/plot/_response_models.py:83]:
```suggestion
        Explanation of the code steps used for `chart_code` field.""",
```
Seems solved the incorrect response type

File: vizro-ai/src/vizro_ai/dashboard/_response_models/components.py (hunk 12 above)
Comment (lingyielia) [reply vizro-ai/src/vizro_ai/dashboard/_response_models/components.py:66]: I think replace the `DebugFailure` with `ValidationError` should do the trick
```
        except ValidationError as e:
            logger.warning(
                f"""
[FALLBACK] Build failed for `Component`, returning default values.
Reason: {e}
Relevant prompt: {self.component_description}
"""

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 14]
```diff
@@ -48,120 +57,42 @@ def __init__(self, model: Optional[Union[ChatOpenAI, str]] = None):
...
 
         """
-        vizro_plot = self._run_plot_tasks(
-            df=df, user_input=user_input, explain=explain, max_debug_retry=max_debug_retry
+        response_model = ChartPlanDynamicFactory(data_frame=df) if validate_code else ChartPlanStatic
+        response = _get_pydantic_model(
+            query=user_input,
+            llm_model=self.model,
+            response_model=response_model,
+            df_info=df,
```
[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py]: We should re-use code, I re-used your code

[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/_vizro_ai.py]: I'm surprised this wasn't in - i had changed that in the past

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 15]
```diff
@@ -48,120 +58,43 @@ def __init__(self, model: Optional[Union[ChatOpenAI, str]] = None):
...
-            max_debug_retry: Maximum number of retries to debug errors. Defaults to `3`.
-            return_elements: Flag to return PlotOutputs dataclass that includes all possible elements generated.
+            max_debug_retry: Maximum number of retries to debug errors. Defaults to `1`.
+            return_elements: Flag to return ChartPlanStatic pydantic model that includes all
+                possible elements generated. Defaults to `False`.
+            validate_code: Flag if produced code should be executed to validate it. Defaults to `True`.
 
         Returns:
-           go.Figure or PlotOutputs dataclass
+            go.Figure or ChartPlanStatic pydantic model
```
Comment (nadijagraca) [inline vizro-ai/src/vizro_ai/_vizro_ai.py:82]: Docstrings should be updated to reflect new changed model name (ChartPlan).

**Review by nadijagraca — APPROVED** [review]
Looks really good. I love seeing so much code deleted, so much cleaner.  🧹 🙌 

I did encounter a small issue with safeguard code validation when trying to run an example. In the plot description I asked for a horizontal line to be added to the chart (the error message is below). 

I'm not sure if this is intended behaviour or not. 
Screenshot 2024-09-05 at 16 02 54

**Review by lingyielia — APPROVED** [review]

**Comment by maxschulz-COL** [comment]
> Looks really good. I love seeing so much code deleted, so much cleaner. 🧹 🙌
> 
> I did encounter a small issue with safeguard code validation when trying to run an example. In the plot description I asked for a horizontal line to be added to the chart (the error message is below).
> 
> I'm not sure if this is intended behaviour or not. Screenshot 2024-09-05 at 16 02 54

Thanks for reporting this - strange, I could not really tell which built-in this is, surely `len` must be allowed haha. But I will check

**Comment by lingyielia** [comment]
> > Looks really good. I love seeing so much code deleted, so much cleaner. 🧹 🙌
> > I did encounter a small issue with safeguard code validation when trying to run an example. In the plot description I asked for a horizontal line to be added to the chart (the error message is below).
> > I'm not sure if this is intended behaviour or not.
> 
> Thanks for reporting this - strange, I could not really tell which built-in this is, surely `len` must be allowed haha. But I will check

Ah it's caused by `type`, which is not a whitelist word

```
raise Exception(
            f"Unsafe builtin functions {builtin_str} are used in generated code line: {code} and cannot be executed. If"
            f" you require a builtin package, reach out to the Vizro team."
        )
```
